observ/watcher.py

```diff
@@ -317,3 +258,54 @@ def depend(self) -> None:
     @property
     def fn_fqn(self) -> str:
         return f"{self.fn.__module__}.{self.fn.__qualname__}"
+
+
+def weak(obj, method):
+    """
+    Prevent the strong capture of the given object by using a weakref.ref instead.
+    Assumes the first argument to the method to be the argument for which a weak ref
+    is made.
+
+    The wrapped method will only be called, if the weak ref object is actually still
+    alive.
+
+    This will ensure that the method does not capture a strong reference to the object.
+    """
+    weak_obj = ref(obj)
+
+    sig = inspect.signature(method)
+    nr_arguments = len(sig.parameters)
+
+    if nr_arguments == 1:
+
+        @wraps(method)
+        def wrapped():
+            if this := weak_obj():
+                return method(this)
+
+        return wrapped
+    elif nr_arguments == 2:
+
+        @wraps(method)
+        def wrapped(new):
+            if this := weak_obj():
+                return method(this, new)
+
+        return wrapped
+    elif nr_arguments == 3:
+
+        @wraps(method)
+        def wrapped(new, old):
+            if this := weak_obj():
+                return method(this, new, old)
+
+        return wrapped
+    else:
+        raise NotImplementedError
```

Comment:
Doesn't this need brackets?

Reply:
Actually, surprisingly, no, it doesn't :) But, your comment made me read the docs for NotImplementedError and I should not be using it here at all! A type error would be better suited.